File: src/main/java/core/procedures/hashmaps/Get.java
package core.procedures.hashmaps;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;

import java.util.Map;

public final class Get extends AFn {

  public Get() {
    super(new FnArgsBuilder().minArgs(2).maxArgs(3).mandatoryArgsTypes(new Class[]{Map.class, Object.class, Object.class}));
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "get";
  }

  @Override
  public Object apply(Object... args) {
    if (args.length == 3) {
      return ((Map) args[0]).getOrDefault(args[1], args[2]);
    } else {
      return ((Map) args[0]).get(args[1]);
    }
  }
}

File: src/main/java/core/procedures/seqs/Get.kt
package core.procedures.seqs

import core.procedures.AFn
import core.utils.Utils

class Get : AFn<Any?, Any?>(name = "get", isPure = true, minArgs = 2, maxArgs = 3) {

    override operator fun invoke(args: Array<out Any?>) = invoke(args[0], args[1], args.getOrNull(2))

    override operator fun invoke(arg1: Any?, arg2: Any?, arg3: Any?) = when {
        arg1 is Map<*, *> -> arg1.getOrDefault(arg2, arg3)
        Utils.isSeqable(arg1) -> when {
            Utils.isExactNonNegativeInteger(arg2) -> Utils.toSequence(arg1).elementAtOrNull((arg2 as Number).toInt()) ?: arg3
            else -> arg2
        }
        else -> arg2
    }
}
